### DISCUSSION TITLE
Using espnet for speech enhancement

### DISCUSSION THREAD

**Discussion by jianupetrut** [espnet/espnet#4060] [Q&A] [answered]
Hello,

I have a beginner question as I'm having a hard time understanding the documentation and was hoping someone would be able to provide a more straightforward answer.

After installing the pip module, what would be the simplest way to use it for speech enhancement (noise reduction) on a local, pre-recorded audio file ?
e.g something like 
`python -m espnet.enh [???] --noisy_file_path=path1 --output_path=path2` from the CLI or some code example that can be turned into a script?

Thank you !

**Comment by sw005320** [comment]
We actually don't have such an independent binary, but we have an API to do some enhancement.
Please check https://colab.research.google.com/drive/1fjRJCh96SoYLZPRxsjF9VDv4Q2VoIckI?usp=sharing

@kamo-naoyuki, is there an easy way to make such a binary?

**Comment by Emrys365** [comment, marked as answer]
You may also refer to this portable script: https://github.com/espnet/espnet/blob/master/egs2/TEMPLATE/enh1/scripts/utils/enhance_dataset.sh#L57
> Instead of enhancing a single audio file, it is designed to enhance a list of audio files with multiprocessing.

An example usage can be found here: https://github.com/espnet/espnet/pull/3487#issue-975855695

**Comment by chenzhehuai** [comment]
Hi, this mail address is unused now. The mail will be forwarded to ***@***.******@***.***

**Comment by jianupetrut** [comment]
Thanks so much for your replies ! The enhance_dataset.sh script seems to be what I need. However, I have an additional question. Running it on a folder containing .wav files I get the following error:

```[enh1]$ ./scripts/utils/enhance_dataset.sh /mnt/efs/fs1/efs/espnetTest/ /mnt/efs/fs1/efs/espnetTestOutput /mnt/efs/fs1/efs/espnet/espnet/egs2/TEMPLATE/enh1/5epoch.pth
./path.sh: line 9: /mnt/efs/fs1/efs/espnet/espnet/egs2/TEMPLATE/enh1/../../../tools/activate_python.sh: No such file or directory
2022-02-11T14:02:02 (enhance_dataset.sh:80:main) ./scripts/utils/enhance_dataset.sh /mnt/efs/fs1/efs/espnetTest/ /mnt/efs/fs1/efs/espnetTestOutput /mnt/efs/fs1/efs/espnet/espnet/egs2/TEMPLATE/enh1/5epoch.pth
2022-02-11T14:02:02 (enhance_dataset.sh:93:main) ./scripts/utils/enhance_dataset.sh: no such file /mnt/efs/fs1/efs/espnetTest//utt2spk
[enh1]$
```
and I've noticed in the enhance_dataset.sh file the following comment : 
```
# This script operates on a directory, such as in data/train/,
# which contains the following files:
#  wav.scp
#  utt2spk
```
What should be in the utt2spk folder and how can I generate this content?

**Comment by Emrys365** [comment]
> Thanks so much for your replies ! The enhance_dataset.sh script seems to be what I need. However, I have an additional question. Running it on a folder containing .wav files I get the following error:
> 
> ```[enh1]$ ./scripts/utils/enhance_dataset.sh /mnt/efs/fs1/efs/espnetTest/ /mnt/efs/fs1/efs/espnetTestOutput /mnt/efs/fs1/efs/espnet/espnet/egs2/TEMPLATE/enh1/5epoch.pth
> ./path.sh: line 9: /mnt/efs/fs1/efs/espnet/espnet/egs2/TEMPLATE/enh1/../../../tools/activate_python.sh: No such file or directory
> 2022-02-11T14:02:02 (enhance_dataset.sh:80:main) ./scripts/utils/enhance_dataset.sh /mnt/efs/fs1/efs/espnetTest/ /mnt/efs/fs1/efs/espnetTestOutput /mnt/efs/fs1/efs/espnet/espnet/egs2/TEMPLATE/enh1/5epoch.pth
> 2022-02-11T14:02:02 (enhance_dataset.sh:93:main) ./scripts/utils/enhance_dataset.sh: no such file /mnt/efs/fs1/efs/espnetTest//utt2spk
> [enh1]$
> ```

The first error in ./path.sh means that you need to install espnet from source to generate activate_python.sh. This may not be necessary and you can ignore it for now.

The second error in enhance_dataset.sh requires a utt2spk file to be placed in the same directory as wav.scp.


> and I've noticed in the enhance_dataset.sh file the following comment : 
> ```
> # This script operates on a directory, such as in data/train/,
> # which contains the following files:
> #  wav.scp
> #  utt2spk
> ```
> What should be in the utt2spk folder and how can I generate this content? 

The format of utt2spk is as follows:
```
utt_id speaker_id
```
For your case, you can simply generate it by using the utt_id as the speaker_id:
```bash
awk '{print $1, $1}' /path/to/wav.scp > /path/to/utt2spk
```

**Comment by jianupetrut** [comment]
Thank you for the explanation!
